### PR TITLE
add community authentication handler

### PR DESCRIPTION
**Pull request by GuillaumeDSM**
requires https://github.com/Drakkar-Software/OctoBot/pull/1387

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -15,3 +15,4 @@ colorlog==4.2.1
 yarl==1.1.0
 idna<2.9,>=2.5
 aiohttp==3.7.2
+requests==2.24.0
```
Comment (Herklos) [inline requirements.txt:18]: Why not using aiohttp ?

File: octobot/community/authentication.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        self.authentication_url = authentication_url
+        self.refresh_token = None
+        self.edited_config = config
+
+        self._auth_token = None
+        self._expire_at = None
+        self._session = requests.Session()
+        self._cache = {}
+
+        if None not in (username, password):
```
Comment (Herklos) [inline octobot/community/authentication.py:52]: I'm not a big fan of `None not in (username, password)`. I would prefer `username and password`

File: octobot/community/authentication.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            return self._cache[url]
+        else:
+            return self._session.get(url, params=params, **kwargs)
+
+    @authenticated
+    def post(self, url, data=None, json=None, **kwargs):
+        return self._session.post(url, data=data, json=json, **kwargs)
+
+    def is_logged_in(self):
+        return None not in (self._auth_token, self.refresh_token, self._expire_at)
```
Comment (Herklos) [inline octobot/community/authentication.py:90]: Same here

File: octobot/community/authentication.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+    def ensure_token_validity(self):
```
Comment (Herklos) [inline octobot/community/authentication.py:92]: :+1:

File: octobot/community/authentication.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            # try to login with saved token if available
+            self._try_auto_login()
+            if not self.is_logged_in():
+                # still not logged in: raise
+                raise AuthenticationRequired()
+        if time.time() >= self._expire_at:
+            self._refresh_auth()
+
+    def remove_login_detail(self):
+        self._update_login_config("")
```
Comment (Herklos) [inline octobot/community/authentication.py:103]: Why not passing no argument with `_update_login_config(self, value="")` ?

File: octobot/community/authentication.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            self._refresh_auth()
+        except FailedAuthentication:
+            self.logout()
+        except Exception as e:
+            self.logger.exception(e, True, f"Error when trying to refresh community login: {e}")
+
+    def _refresh_auth(self):
+        params = {
+            "refresh_token": self.refresh_token,
+            "grant_type": "refresh_token",
```
Comment (Herklos) [inline octobot/community/authentication.py:130]: Can we add a class constant for this two strings ?

File: tests/unit_tests/community/test_authentication.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+import pytest
+import mock
+import requests
+import time
+
+import octobot.community as community
+import octobot_commons.config_manager
+import octobot_commons.constants as commons_constants
+
+AUTH_URL = "https://oh.fake/auth"
```
Comment (Herklos) [inline tests/unit_tests/community/test_authentication.py:25]: Could we use `os.getenv("COMMUNITY_SERVER_URL", "https://oh.fake/auth")` to test with a real server ? (same for www.todo.com below)

File: tests/unit_tests/community/test_authentication.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+            mock.patch.object(community.CommunityAuthentication, "_handle_auth_result", mock.Mock()) \
+            as auth_res_mock, \
+            mock.patch.object(requests, "post", mock.Mock(return_value="1")) as post_mock:
+        auth.login("username", "password")
+        reset_mock.assert_called_once()
+        post_mock.assert_called_once()
+        auth_res_mock.assert_called_once_with("1")
+
+
+def test_logout(auth):
```
Comment (Herklos) [inline tests/unit_tests/community/test_authentication.py:81]: :+1:

File: tests/unit_tests/community/test_authentication.py [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+    assert auth.is_logged_in() is False
+    auth._auth_token = "1"
+    auth.refresh_token = "1"
+    auth._expire_at = None
+    assert auth.is_logged_in() is False
+    auth._expire_at = "1"
+    assert auth.is_logged_in() is True
+
+
+def test_ensure_token_validity():
```
Comment (Herklos) [inline tests/unit_tests/community/test_authentication.py:216]: :100:

**Review by Herklos — COMMENTED** [review]
Very nice job :+1: Great test coverage :ok_hand:

File: octobot/community/authentication.py (hunk 2 above)
Comment (GuillaumeDSM) [reply octobot/community/authentication.py:52]: The problem with this check is that it prevents the opportunity for an empty password / username auth

Comment (Herklos) [reply octobot/community/authentication.py:52]: Do you think we should able to use empty login username and password ?

File: octobot/community/authentication.py (hunk 5 above)
Comment (GuillaumeDSM) [reply octobot/community/authentication.py:103]: Here `_update_login_config` is supposed to receive a value, "" is more of an exception than the default behavior

Comment (Herklos) [reply octobot/community/authentication.py:103]: Ok, in my opinion it's not very clear to say "update something with an empty value" to reset / remove a thing.

File: octobot/community/authentication.py (hunk 2 above)
Comment (GuillaumeDSM) [reply octobot/community/authentication.py:52]: I just don't see why we should forbid

File: octobot/community/authentication.py (hunk 5 above)
Comment (GuillaumeDSM) [reply octobot/community/authentication.py:103]: Right, I'll see how to make it clearer

**Review by Herklos — APPROVED** [review]

**Review by Herklos — APPROVED** [review]

**Comment by GuillaumeDSM** [comment]
contained in https://github.com/Drakkar-Software/OctoBot/pull/1395

**Pull request closed without merging**